### PR TITLE
Added new widget

### PR DESCRIPTION
**Pull request by ar4s**
New widget replaced standard widget for fields in `raw_id_fields`
(`ForeignKeyRawIdWidget`).

Connected to #1613

### CODE REVIEW

File: src/ralph/admin/suggest.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+            'results': [
+                {'pk': obj.pk, '__str__': str(obj)}
+                for obj in self.get_queryset()
+            ]
+        })
+
+
+class AjaxSuggestMixin(object):
+    def get_urls(self):
+        from django.conf.urls import url
```
Comment (ghost) [inline src/ralph/admin/suggest.py]: why here? + it looks like it is unused

File: src/ralph/admin/suggest.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    """
+    A mixin that can be used to render a JSON response.
+    """
+    def render_to_json_response(self, context, **response_kwargs):
+        """
+        Returns a JSON response, transforming 'context' to make the payload.
+        """
+        return JsonResponse(self.get_data(context), **response_kwargs)
+
+    def get_data(self, context):
```
Comment (ghost) [inline src/ralph/admin/suggest.py]: Is this needed?

File: src/ralph/admin/suggest.py (hunk 1 above)
Comment (ar4s) [inline src/ralph/admin/suggest.py]: it used here https://github.com/allegro/ralph/pull/1627/files#diff-73c1233d9acef15d6945e388915ad4f4R71

I will move this import block to top.

File: src/ralph/admin/suggest.py (hunk 2 above)
Comment (ar4s) [inline src/ralph/admin/suggest.py]: nope :smile:

File: src/ralph/admin/widgets.py [hunk 3]
```diff
@@ -114,3 +119,63 @@ def render_options(self, choices, selected_choices):
...
+    def media(self):
+        js = map(lambda x: os.path.join(*x), [
+            ('src', 'js', 'ralph-suggest.js'),
+        ])
+        return forms.Media(js=[
+            static(path) for path in js
+        ])
+
+    def get_url_parameters(self):
+        from django.contrib.admin.views.main import TO_FIELD_VAR
```
Comment (ghost) [inline src/ralph/admin/widgets.py]: top of file?

Comment (ar4s) [inline src/ralph/admin/widgets.py]: ok

File: src/ralph/admin/widgets.py [hunk 4]
```diff
@@ -114,3 +119,63 @@ def render_options(self, choices, selected_choices):
...
+class SuggestWidget(forms.TextInput):
+    def __init__(self, rel, admin_site, attrs=None, using=None):
+        self.rel = rel
+        self.admin_site = admin_site
+        self.db = using
+        super().__init__(attrs)
+
+    @property
+    def media(self):
+        js = map(lambda x: os.path.join(*x), [
```
Comment (ghost) [inline src/ralph/admin/widgets.py]: what's the purpose of this?

File: src/ralph/admin/widgets.py [hunk 5]
```diff
@@ -114,3 +119,63 @@ def render_options(self, choices, selected_choices):
...
+        return forms.Media(js=[
+            static(path) for path in js
+        ])
+
+    def get_url_parameters(self):
+        from django.contrib.admin.views.main import TO_FIELD_VAR
+        params = {
+            TO_FIELD_VAR: self.rel.get_related_field().name
+        }
+        return '?' + '&amp;'.join('%s=%s' % (k, v) for k, v in params.items())
```
Comment (ghost) [inline src/ralph/admin/widgets.py]: https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlencode

```
>>> from urllib import parse
>>> dd = {'a': 1, 'b':2}
>>> parse.urlencode(dd)
'b=2&a=1'
```

File: src/ralph/admin/suggest.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    A mixin that can be used to render a JSON response.
+    """
+    def render_to_json_response(self, context, **response_kwargs):
+        """
+        Returns a JSON response, transforming 'context' to make the payload.
+        """
+        return JsonResponse(context, **response_kwargs)
+
+
+class SuggestView(JsonViewMixin, View):
```
Comment (mkurek) [inline src/ralph/admin/suggest.py]: add (simple & short) docstring

File: src/ralph/admin/suggest.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        """
+        return self.render_to_json_response({
+            'results': [
+                {'pk': obj.pk, '__str__': str(obj)}
+                for obj in self.get_queryset()
+            ]
+        })
+
+
+class AjaxSuggestMixin(object):
```
Comment (mkurek) [inline src/ralph/admin/suggest.py]: add (simple & short) docstring

Comment (ar4s) [inline src/ralph/admin/suggest.py]: done

File: src/ralph/admin/suggest.py (hunk 6 above)
Comment (ar4s) [inline src/ralph/admin/suggest.py]: done

**Comment by ghost** [comment]
:ok:

File: src/ralph/admin/autocomplete.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+            def dispatch(self, request, *args, **kwargs):
+                pk = request.GET.get(DETAIL_PARAM, None)
+                if not pk:
+                    return HttpResponseBadRequest()
+                self.object = get_object_or_404(model, pk=int(pk))
+                return super().dispatch(request, *args, **kwargs)
+
+            def get_queryset(self):
+                return [self.object]
```
Comment (mkurek) [inline src/ralph/admin/autocomplete.py]: shouldn't this method return `QuerySet` instance?

File: src/ralph/static/src/js/ralph-autocomplete.js [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+            set: function(val) {
+                that.changeTargetAfterPopup(val);
+                this.setAttribute('value', val);
+            }
+        });
+
+    };
+    AutocompleteWidget.prototype.keyUp = function(event) {
+        var that = this;
+        var regex = new RegExp('^[a-zA-Z0-9]+$');
```
Comment (mkurek) [inline src/ralph/static/src/js/ralph-autocomplete.js]: fix this regexp (to accept space, - etc)

File: src/ralph/static/src/js/ralph-autocomplete.js [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+        var data = {}
+        data[that.options.detailVar] = val;
+        that.fetch(that.options.detailsUrl, data, function(data) {
+            that.editMode(false);
+            $('.title', that.$currentItem).html(data.results[0].__str__);
+        });
+
+    };
+    AutocompleteWidget.prototype.editMode = function(on) {
+        var that = this;
```
Comment (mkurek) [inline src/ralph/static/src/js/ralph-autocomplete.js]: this is not necessary here.

File: src/ralph/admin/templates/admin/widgets/suggest.html [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+<div class="suggest-field-widget{% if not current_object %} edit{% endif %}"{{ attrs }}>
+    <div class="widget-wrapper">
+        <div class="item-wrapper">
+            <div class="item">
+                <span class="title">{{ current_object }}</span>
+                <i class="fa fa-times delete"></i>
+            </div>
+        </div>
+        <input class="input" type="text" placeholder="{% trans 'Please type some sentence..' %}"/>
```
Comment (mkurek) [inline src/ralph/admin/templates/admin/widgets/suggest.html]: use text from search box, ex. 'Search by <field 1>, <field 2>, ...`

**Comment by mkurek** [comment]
:ok:

**Pull request merged**